server/src/main/java/org/opensearch/search/aggregations/NonGlobalAggCollectorManager.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.aggregations;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+import org.opensearch.search.internal.SearchContext;
+import org.opensearch.search.profile.query.CollectorResult;
+
+import java.io.IOException;
+
+/**
+ * {@link CollectorManager} to take care of non-global aggregation operators in case of concurrent segment search
+ */
+public class NonGlobalAggCollectorManager extends AggregationCollectorManager {
+
+    private Collector collector;
+
+    public NonGlobalAggCollectorManager(SearchContext context) throws IOException {
+        super(context, context.aggregations().factories()::createTopLevelNonGlobalAggregators, CollectorResult.REASON_AGGREGATION);
+        collector = super.newCollector();
+    }
+
+    @Override
+    public Collector newCollector() throws IOException {
+        if (collector != null) {
```

Comment:
My apologies, I don't understand this code: the `collector` is not used anywhere, why we're creating? To me, it seems like the `NonGlobalAggCollectorManager` is not needed at all and `AggregationCollectorManager` could be used instead.

Reply:
In the `AggregationProcessor::preProcess` it creates the collector instances for aggregation and throws `AggregationInitializationException` if there is any exception during agg collector creation. To throw the same error and catch initialization issues with Aggregation, this `collector` is initialized in constructor of `NonGlobalAggCollectorManager` which is used in concurrent search path. If we use `newCollector` as is instead then the aggregation initialization issues will be caught later during the execution and the exception thrown to the client will a generic one wrapped inside `QueryPhaseExecutionException`

Reply:
Haven't looked but what is wrong with throwing `QueryPhaseExecutionException`? And the code still calls the `super.newCollector()` so the `QueryPhaseExecutionException` may be thrown anyway?

Reply:
Keeping the error explicit that issue is with aggregation initialization for better error handling. If the first `collector` creation goes fine then generally will expect the `super.newCollector` to not have initialization issues as well.

Reply:
I don't think this is worth any benefits if `QueryPhaseExecutionException ` does the job - we would have the cause available in any case

Reply:
The problem of letting failure handling happen during `newCollector` call becomes bad with global aggregation case.  In case of global aggregation the exception will happen when `newCollector` is called  in `postProcess` which is after the query execution is completed. To follow the fail fast strategy I preferred to initialize the collectors upfront even though the execution happens later in `postProcess`. So still see having this implementation is useful in such cases to throw error before spending resources to execute the search request. Also I think this way it will be consistent with the non concurrent path which also throws the `AggregationInitializationException` with any issues for aggregation collectors before performing the query execution.

Reply:
I see your point, but it becomes very inconsistent with concurrent path where the exception will always be raised in `newCollector()`, plus we both agree (looking on implementation) that because of the delegation to the super.newCollector() there are absolutely no guarantee that no exception would be raised. I would certainly prefer to rely on uniform behavior between concurrent and non-concurrent path, otherwise the same exception will be reported very differently.

Reply:
> but it becomes very inconsistent with concurrent path where the exception will always be raised in newCollector(), 

Not sure if I understand this

>plus we both agree (looking on implementation) that because of the delegation to the super.newCollector() there are absolutely no guarantee that no exception would be raised.

However I don't see a reason why exception will be raised in subsequent call but not in first call since the setup is independent of number of times (or slices) it is being called for.

Reply:
> Not sure if I understand this:

Let say the `AggregationCollectorManager::newCollector() implementation` has a bung and throws `NullPointerException` for whatever reason (not important), for concurrent flow - the exception will be reported as `QueryPhaseExecutionException`, but non-concurrent as `AggregationInitializationException`, why? 

> However I don't see a reason why exception will be raised in subsequent call but not in first call since the setup is independent of number of times (or slices) it is being called for.

If it can happen, it will happen.

Reply:
In concurrent case as well during `preProcess` it will [create the global/non-global](https://github.com/opensearch-project/OpenSearch/pull/7514/files#diff-28d65e7be4bf91608a1b02e9745afc3765236b2484fe6b4da487440500930fadR38) `AggregationCollectorManager` which will call `super.newCollector` in constructor. So if there is any bug in `newCollector` method that will throw exception and behavior will be same as non concurrent path.

>> If it can happen, it will happen.

I am not sure in what case it can happen such that first call succeeds and next call fails. If it does happen like that then probably there is some bug which needs to be handled

Reply:
Oh I see now (sorry I missed the big picture), all "final" collectors follow the same pattern, it generally makes sense now, please make this code thread safe (the `newCollector` will be called by multiple threads, the `collector` assignment to `null` must be visible to all threads).

Reply:
> please make this code thread safe (the newCollector will be called by multiple threads, the collector assignment to null must be visible to all threads).

This is not needed as `newCollector` is called on the [same thread](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L665) (which is the calling thread) and then later each `collector` instance is handed over to the slice executor threads

Reply:
:100: ACK, thank you